File: sections/40_data_structures_mixed_contest/problems/b_list_removals/solution.cpp
#include <bits/stdc++.h>
using namespace std;

struct Fenwick {
    int n;
    vector<int> bit;
    Fenwick(int n) : n(n), bit(n + 1, 0) {}
    void add(int i, int delta) {
        for (; i <= n; i += i & -i) bit[i] += delta;
    }
    int kth(int k) const {
        int pos = 0;
        int step = 1;
        while ((step << 1) <= n) step <<= 1;
        for (; step; step >>= 1) {
            int nxt = pos + step;
            if (nxt <= n && bit[nxt] < k) {
                pos = nxt;
                k -= bit[nxt];
            }
        }
        return pos + 1;
    }
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    int n;
    if (!(cin >> n)) return 0;
    vector<long long> a(n + 1);
    for (int i = 1; i <= n; i++) cin >> a[i];
    Fenwick bit(n);
    for (int i = 1; i <= n; i++) bit.add(i, 1);
    for (int t = 0; t < n; t++) {
        int k;
        cin >> k;
        int idx = bit.kth(k);
        if (t) cout << ' ';
        cout << a[idx];
        bit.add(idx, -1);
    }
    cout << '\n';
    return 0;
}

File: sections/40_data_structures_mixed_contest/problems/b_list_removals/solution.py
import sys

class Fenwick:

    def __init__(self, n):
        self.n = n
        self.bit = [0] * (n + 1)

    def add(self, i, delta):
        while i <= self.n:
            self.bit[i] += delta
            i += i & -i

    def kth(self, k):
        pos = 0
        bit = 1 << self.n.bit_length()
        while bit:
            nxt = pos + bit
            if nxt <= self.n and self.bit[nxt] < k:
                pos = nxt
                k -= self.bit[nxt]
            bit //= 2
        return pos + 1

def main():
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return
    n = data[0]
    a = data[1:1 + n]
    positions = data[1 + n:1 + 2 * n]
    bit = Fenwick(n)
    for i in range(1, n + 1):
        bit.add(i, 1)
    out = []
    for k in positions:
        idx = bit.kth(k)
        out.append(str(a[idx - 1]))
        bit.add(idx, -1)
    print(' '.join(out))
if __name__ == '__main__':
    main()
